lib/token.h

```diff
@@ -789,6 +806,30 @@ class CPPCHECKLIB Token {
     static Token *findmatch(Token * const startTok, const char pattern[], const Token * const end, const nonneg int varId = 0);
 
 private:
+    template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*> )>
+    static T *tokAtImpl(T *tok, int index)
+    {
+        while (index > 0 && tok) {
+            tok = tok->next();
+            --index;
+        }
+        while (index < 0 && tok) {
+            tok = tok->previous();
+            ++index;
+        }
+        return tok;
+    }
+
+    template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*> )>
+    static T *linkAtImpl(T *thisTok, int index)
+    {
+        T *tok = thisTok->tokAt(index);
+        if (!tok) {
+            throw InternalError(thisTok, "Internal error. Token::linkAt called with index outside the tokens range.");
```

Comment:
I wonder if we should remove this and just let the code crash.

Reply:
Does it affect performance?

Reply:
Technically it should because it can raise an exception (hence why `noexcept` is a thing).

In a short test:
Clang 17 `1,194,021,239` -> `1,191,122,393`